atom-client/src/main/java/org/motechproject/atomclient/service/impl/FeedCache.java

```diff
@@ -0,0 +1,290 @@
+import org.motechproject.atomclient.repository.FeedRecordDataService;
+import org.motechproject.atomclient.service.AtomClientConfigService;
+import org.motechproject.atomclient.service.Constants;
+import org.motechproject.event.MotechEvent;
+import org.motechproject.event.listener.EventRelay;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class FeedCache implements FeedFetcherCache {
+
+
+    public static final String FEED_DATA = "feedData";
+    private FeedRecordDataService feedRecordDataService;
+    private AtomClientConfigService atomClientConfigService;
+    private EventRelay eventRelay;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FeedCache.class);
+
+
+    public FeedCache(FeedRecordDataService feedRecordDataService, EventRelay eventRelay,
+                     AtomClientConfigService atomClientConfigService) {
+        this.feedRecordDataService = feedRecordDataService;
+        this.eventRelay = eventRelay;
+        this.atomClientConfigService = atomClientConfigService;
+    }
+
+
+    public static String feedToString(SyndFeed feed) throws FeedException, UnsupportedEncodingException {
+        SyndFeedOutput syndFeedOutput = new SyndFeedOutput();
+        return syndFeedOutput.outputString(feed);
+    }
+
+
+    public static SyndFeed feedFromString(String xml) throws IOException, ClassNotFoundException, FeedException {
+        SyndFeedInput syndFeedInput = new SyndFeedInput();
+        return syndFeedInput.build(new StringReader(xml));
+    }
+
+
+    public static String urlToString(URL url) throws IOException {
+        return url.toExternalForm();
+    }
+
+
+    public static URL urlFromString(String s) throws IOException {
+        return new URL(s);
+    }
+
+
+    public SyndFeedInfo feedRecordToFeedInfo(FeedRecord record) throws IOException, ClassNotFoundException, FeedException {
+        SyndFeedInfo info = new SyndFeedInfo();
+        info.setUrl(urlFromString(record.getUrl()));
+        info.setLastModified(record.getLastModified());
+        info.setSyndFeed(feedFromString(record.getData()));
+        return info;
+    }
+
+
+    /**
+     * Get a SyndFeedInfo object from the cache.
+     *
+     * @param feedUrl The url of the feed
+     * @return A SyndFeedInfo or null if it is not in the cache
+     */
+    @Override
+    public SyndFeedInfo getFeedInfo(URL feedUrl) {
+        try {
+            String url = urlToString(feedUrl);
+            FeedRecord record = feedRecordDataService.findByURL(url);
+            if (record != null) {
+                return feedRecordToFeedInfo(record);
+            }
+        } catch (IOException | ClassNotFoundException | FeedException ex) {
+            LOGGER.error(ex.getMessage(), ex);
```

Comment:
Can we additionally add our own message, to signal where the problem occured? Eg. "Error while retrieving object from cache".